### PR TITLE
Enable performance-unnecessary-value-param in .clang-tidy

### PR DESCRIPTION
**Pull request by goldsborough**
This PR fixes around 250 places in the codebase where we were making unnecessary copies of objects (some large, some small).

@ezyang

### CODE REVIEW

File: torch/csrc/jit/ir.cpp [hunk 1]
```diff
@@ -1294,7 +1295,7 @@ Value* Graph::insert(
     Symbol opname,
```
Comment (zdevito) [inline torch/csrc/jit/ir.cpp]: Why is it ok to pass Symbol by value but not other types? What feature of a type causes this error to occur? What if we change Symbol to no longer have this feature?

**Review by zdevito — COMMENTED** [review]
I do think it is a good idea to make the suggested changes for truly big objects (in this PR, std::string, std::vector, and SourceRange), and it would be great if we could apply this to specific types (vector, map, string), but it seems too pedantic to apply this across the board.

For Expr, Tensor, optional<Device>, I doubt these changes meaningfully improved performance, but they certainly make the code less readable (notice how many times we end up having to split a function signature because of the extra noise in the declaration). It is not clear whether in the future we will end up having more objects like Device for which these changes are actually a de-optimization.

File: torch/csrc/jit/ir.cpp (hunk 1 above)
Comment (goldsborough) [reply torch/csrc/jit/ir.cpp]: Because clant-tidy detects that it is a trivial type (`uint32_t`)

**Comment by goldsborough** [comment]
The way I see these lint checks is that the fixes are distributed as follows:
- 10% very important bug fixes / optimizations,
- 80% a good idea,
- 10% wouldn't normally do it, but worth the extra effort for the above 90%

I think you have to ask yourself whether you're willing to put in more effort 10% of the time in order while getting legit benefits 90% of the time. I agree that for `Device` and even `Tensor` it'd be fine not use `const&` (although `Tensor` we're bumping atomic refcounts again), but I just can't make `clang-tidy` that fine grained and personally I prefer writing clean/good practice code all the time. Just to look at all the actual fixes that are in this PR:

- `std::string` taken by value all over the place,
- `std::function` taken by value all over the place, which means copying things when a closure stores things,
- lots of `const` without `&`, like `const std::unordered_map<std::string, PyObject*> kwargs`, which just looks like an unintended performance bug to me (the person probably wanted `const&`),
- `shared_ptr`s being copied like they're ints.

I would also consider the fact that lots of new developers on the team are new to C++ and don't fully get the value semantics of C++. In that sense a linter is actually educational because it teaches you to not pass a `std::string` by value which new people might not know is a bad thing.

That said, I understand different folks have different priorities, and if you guys still feel like the extra `const&` and `std::move`s will greatly harm productivity, I can just undo the changes in `torch/csrc/jit`, disable the clang-tidy check and only check in the changes to other parts of the codebase.

File: torch/csrc/jit/ir.cpp (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/ir.cpp]: Symbol is a struct. Is it checking a certain property on the struct?

**Comment by zdevito** [comment]
I am wary of this, but I am willing to let it run for a couple weeks and see how it goes. My concern is about how much effort it will take to continually appease the pass and trying it will quantify that. If it turns out not to be a big burden then I suppose the quirks are not horrible.

**Review by zdevito — APPROVED** [review]

**Pull request closed without merging**